Source/Core/Core/IOS/FS/HostBackend/FS.cpp

```diff
@@ -252,99 +255,135 @@ HostFileSystem::FstEntry* HostFileSystem::GetFstEntryForPath(const std::string&
+      u32 count = size;
+      while (count > BUFFER_CHUNK_SIZE)
       {
-        u32 size = 0;
-        p.Do(size);
-
-        File::IOFile handle(name, "wb");
-        char buf[65536];
-        u32 count = size;
-        while (count > 65536)
-        {
-          p.DoArray(buf);
-          handle.WriteArray(&buf[0], 65536);
-          count -= 65536;
-        }
-        p.DoArray(&buf[0], count);
-        handle.WriteArray(&buf[0], count);
-        break;
-      }
+        p.DoArray(buf);
+        handle.WriteArray(&buf[0], BUFFER_CHUNK_SIZE);
+        count -= BUFFER_CHUNK_SIZE;
       }
+      p.DoArray(&buf[0], count);
+      handle.WriteArray(&buf[0], count);
+      break;
+    }
     }
   }
-  else
-  {
-    // recurse through tmp and save dirs and files
+}
 
-    File::FSTEntry parent_entry = File::ScanDirectoryTree(Path, true);
-    std::deque<File::FSTEntry> todo;
-    todo.insert(todo.end(), parent_entry.children.begin(), parent_entry.children.end());
+void HostFileSystem::DoStateWriteOrMeasure(PointerWrap& p, std::string startDirectoryPath)
+{
+  std::string Path = BuildFilename(startDirectoryPath).host_path;
+  File::FSTEntry parent_entry = File::ScanDirectoryTree(Path, true);
+  std::deque<File::FSTEntry> todo;
+  todo.insert(todo.end(), parent_entry.children.begin(), parent_entry.children.end());
 
-    while (!todo.empty())
+  while (!todo.empty())
+  {
+    File::FSTEntry& entry = todo.front();
+    std::string name = entry.physicalName;
+    name.erase(0, Path.length() + 1);
+    char type = entry.isDirectory ? 'd' : 'f';
+    p.Do(type);
+    p.Do(name);
+    if (entry.isDirectory)
+    {
+      todo.insert(todo.end(), entry.children.begin(), entry.children.end());
+    }
+    else
     {
-      File::FSTEntry& entry = todo.front();
-      std::string name = entry.physicalName;
-      name.erase(0, Path.length() + 1);
-      char type = entry.isDirectory ? 'd' : 'f';
-      p.Do(type);
-      p.Do(name);
-      if (entry.isDirectory)
+      u32 size = (u32)entry.size;
+      p.Do(size);
+
+      File::IOFile handle(entry.physicalName, "rb");
+      char buf[BUFFER_CHUNK_SIZE];
+      u32 count = size;
+      while (count > BUFFER_CHUNK_SIZE)
       {
-        todo.insert(todo.end(), entry.children.begin(), entry.children.end());
+        handle.ReadArray(&buf[0], BUFFER_CHUNK_SIZE);
+        p.DoArray(buf);
+        count -= BUFFER_CHUNK_SIZE;
       }
-      else
+      handle.ReadArray(&buf[0], count);
+      p.DoArray(&buf[0], count);
+    }
+    todo.pop_front();
+  }
+
+  char type = 0;
+  p.Do(type);
+}
+
+void HostFileSystem::DoState(PointerWrap& p)
+{
+  // Temporarily close the file, to prevent any issues with the savestating of files/folders.
+  for (Handle& handle : m_handles)
+    handle.host_file.reset();
+
+  bool originalSaveStateMadeDuringMovieRecording = Movie::IsMovieActive();
+  p.Do(originalSaveStateMadeDuringMovieRecording);
+
+  u32 tempVal = 0;
+
+  if (!p.IsReadMode())
```

Comment:
There should be a comment somewhere around here explaining what exactly is happening with the NAND + Movie savestating, with it storing the size of the NAND data so it can skip it on load if required. I understand it right now but I'm not sure how long it would take me to re-understand it in a year or whatever.